# Exercise 2
Add the specified code for each code cell, running the cells _in order_.

Open the included **`data/words_small.txt`** file (containing a "small" list of English words) and `read()` it into a single string variable **`words_str`**. Then print the first 100 _characters_ of the `words_str` variable (use a slice operator!)

In [1]:
words_str = 'a'
with open('data/words_small.txt') as my_file:
    words_str = my_file.read()
print(words_str[0:101])

abbreviate
abbreviation
abdomen
able
aboard
abroad
absent
absolute
abuse
academic
academy
accelerate



**Using a loop**, count the number of times the newline character `\n` appears in the `words_str`. This should tell you the number of words in the file (plus or minus one). You can check your work with the built-in `count()` string method (but use the loop for practice!)

In [3]:
count = 0
for character in words_str:
    if character == '\n':
        count += 1
print(count)
print(words_str.count('\n'))

1520
1520


Create a variable **`letter_counts`** that is a dictionary whose keys are characters (letters) and whose values are the number of times that character appears in `words_str`. Print the dictionary after you make it.
- To do this, start with an empty dictionary, then loop through the letters one at a time. At each letter, increase the value associated with that letter key by one.
- Watch out for if the letter is not in the dictionary yet!
- Also _delete_ the entry for the newline `\n` character

In [6]:
letter_counts = { }
for character in words_str:
    if character in letter_counts.keys():
        letter_counts[character] = 1 + letter_counts[character]
    else:
        letter_counts[character] = 1
print(letter_counts)

#for letter in words_str:
    #previous_count= letter_counts.get(letter, 0)
    #letter_counts[letter] = previous_count + 1
            

{'a': 703, 'b': 188, 'r': 642, 'e': 1118, 'v': 85, 'i': 515, 't': 604, '\n': 1520, 'o': 598, 'n': 550, 'd': 314, 'm': 223, 'l': 457, 's': 504, 'u': 279, 'c': 328, 'y': 189, 'p': 224, 'h': 295, 'f': 155, 'w': 161, 'g': 213, 'k': 126, 'z': 15, 'x': 23, 'j': 16, 'q': 9}


Print out all the count in the `letter_counts` dictionary in the format `"A appears 703 times"`, each on a new line, in alphabetical order. _Hint_: iterate through the sorted keys!

In [12]:
sorted_letter_count = sorted(letter_counts.keys())
for key in sorted_letter_count:
    print(key, " appears ", letter_counts[key],  " times")


  appears  1520  times
a  appears  703  times
b  appears  188  times
c  appears  328  times
d  appears  314  times
e  appears  1118  times
f  appears  155  times
g  appears  213  times
h  appears  295  times
i  appears  515  times
j  appears  16  times
k  appears  126  times
l  appears  457  times
m  appears  223  times
n  appears  550  times
o  appears  598  times
p  appears  224  times
q  appears  9  times
r  appears  642  times
s  appears  504  times
t  appears  604  times
u  appears  279  times
v  appears  85  times
w  appears  161  times
x  appears  23  times
y  appears  189  times
z  appears  15  times


Print which letter from the `letter_counts` dictionary has a **value** of `515`. _Hint:_ use a loop to look through all the _items_ until you find the right one, then print that!

In [15]:
for letter in letter_counts:
    if letter_counts[letter] == 515:
        print(letter.upper())

I


Create a variable **`count_to_letter_dict`** that is a dictionary whose **keys** are the counts and whose **values** are the letters that have that count. This is a _reverse lookup_ dictionary. Hint: ceate a new dictionary, then loop through items in `letter_counts` and add an item to your new dictionary for each element. You can assume no duplicated counts.

In [16]:
count_to_letter_dict = {}
for letter in letter_counts:
     count_to_letter_dict[letter_counts[letter]] = letter 
print(count_to_letter_dict)

{703: 'a', 188: 'b', 642: 'r', 1118: 'e', 85: 'v', 515: 'i', 604: 't', 1520: '\n', 598: 'o', 550: 'n', 314: 'd', 223: 'm', 457: 'l', 504: 's', 279: 'u', 328: 'c', 189: 'y', 224: 'p', 295: 'h', 155: 'f', 161: 'w', 213: 'g', 126: 'k', 15: 'z', 23: 'x', 16: 'j', 9: 'q'}


Print the **10** most common letters and their counts with the format `"A appears 703 times"`. _Hint:_ you'll need to sort the keys to `count_to_letter_dict` and take the first 10 of those.

In [17]:
sorted_count_to_letter = sorted(count_to_letter_dict, reverse = True)
i = 0
while i < 10:
    print( count_to_letter_dict[sorted_count_to_letter[i]], "appears", sorted_count_to_letter[i],  "times." )
    i = i + 1


 appears 1520 times.
e appears 1118 times.
a appears 703 times.
r appears 642 times.
t appears 604 times.
o appears 598 times.
n appears 550 times.
i appears 515 times.
s appears 504 times.
l appears 457 times.


Bonus: repeat your analysis on the `data/words_huge.txt` file! Are the top letters the same?